### ISSUE TITLE
Does NeMo support dynamic YaRN?

### ISSUE THREAD

**Issue by tjoymeed** [NVIDIA/NeMo#13737]
Hi Team,

Thanks for your excellent work!

Does NeMo support dynamic YaRN?

Thanks

**Comment by BoxiangW** [comment]
Hi, thanks for the comment, I think it is not yet supported. We will try to add it in next release maybe.

**Comment by tjoymeed** [comment]
So could you please make a list about:

What features of Qwen 2.5 are fully supported?
What features of Qwen 2.5 are not supported?

I am very confused and couldn't find such a list.

So I am not sure after spending a lot of computes and money, the resultant model is the right one or not.

Thanks!

**Comment by BoxiangW** [comment]
@suiyoubi Do you think you have a list of Qwen 2.5 related features?

**Comment by suiyoubi** [comment]
From what I understand NeMo does not support YaRN, and the current implementation of NeMo Qwen2.5 is based on the HuggingFace model its original config (i.e. for the 32B model https://huggingface.co/Qwen/Qwen2.5-32B-Instruct/blob/main/config.json ), which does not use rope scaling.

The model config for Qwen2.5 is listed here: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/gpt/model/qwen2.py#L123 , and we treat Qwen2.5 indifferently as Qwen2.

**Comment by tjoymeed** [comment]
Thanks!
In order to nail down the details, lets first take Qwen 2.5 7B base model and only consider pretraining 1st stage(i.e. the resultant model context length is 4096):
What features of Qwen 2.5 7B are fully supported?
What features of Qwen 2.5 7B are not supported?

**Comment by suiyoubi** [comment]
I am still a bit confused about "features" here, do you mean pretrain/finetune/lora training feature or low-level feature like Parallelism (TP/PP/CP/etc)

**Comment by tjoymeed** [comment]
I meant replicating the Qwen 2.5 architectural features and pretraining phases.
i.e. after spending time and money into the pretraining using NeMo, what will we get in the resultant model? is that an authentic Qwen 2.5 model(of course the data is own data)

**Comment by suiyoubi** [comment]
Qwen2.5 architectural features are supported. We verified the weights conversion and inference results are identical across NeMo FW and HuggingFace. (i.e. Model FWD)
We also verified for the finetuning stage we can reach comparable results with other models.

For pretraining, I don't know how to answer your question of _what will we get in the resultant model? is that an authentic Qwen 2.5 model(of course the data is own data)_. If you own the Qwen2.5 original data, after the pretraining 4K stage you should reach the same model as the original Qwen2.5 model (without long context adaptation), but if your data is different, I don't think the accuracy of the model will be identical to the original model (and I believe this is expected)

**Comment by tjoymeed** [comment]
Yes, of course, with own data, it won't be 100% Qwen 2.5. 

So aknowledging that, we are talking about:

[Lets take Qwen 2.5 7B] 

(1) architectural characteristics

Rotary Position Embedding (RoPE): Enhances positional encoding for improved long-context understanding.

SwiGLU Activation Functions: Combines Swish and Gated Linear Unit (GLU) for better gradient flow and model expressiveness.

RMSNorm Layer Normalization: Replaces standard layer normalization for training stability and computational efficiency.

Attention QKV Bias: Introduces learnable biases in query, key, and value projections to refine attention patterns.

Dynamic Yarn

DCA

etc.

(2) pretraining characteristics:

Two phases, first 4096, 

Then 32768

---------------------------

What features of Qwen 2.5 7B are currently fully supported?
What features of Qwen 2.5 7B are currently not supported?

**Comment by suiyoubi** [comment]
(1) architectural characteristics
Rotary Position Embedding (RoPE) & SwiGLU Activation Functions & RMSNorm Layer Normalization &Attention QKV Bias these are all supported in NeMo and used for Qwen models.
Long context related config (Yarn specific) are not supported at this point. We support Sliding Window Attention for Long Context adapation.
(2) pretraining characteristics:
First stage pretraining with 4K seq length is supported. Second stage long context adaptation we don't have the exact long context method as Qwen2.5 has.

Hope this help.

**Comment by tjoymeed** [comment]
oaky thanks a lot!
Now I understood.
So first stage pretraining is 100% supported.
But second stage pretraining is not 100% supported yet.
Could you please make the second stage available?
Thanks a lot!

**Comment by suiyoubi** [comment]
Sounds good. I will relay your request for YaRN implementation to the team and prioritize accordingly. Thanks!

**Comment by tjoymeed** [comment]
Great! Thanks a lot! That's very important for us - we are using your tech stack and tool chain in our development - pretraining from scratch. 

Our dollar investment rely on your tech stack and tool chain.

Thanks a lot again!

**Comment by tjoymeed** [comment]
Also the second phase include other techniques such as DCA, ABF, NTK-Aware Scaling etc. am I right?

**Comment by suiyoubi** [comment]
I don't think DCA is something on the plan yet. @BoxiangW  mentioned that YaRN might be able to squeeze in the next release. 

I am not awared of qwen uses ABF, NTK-aware scaling. From what I understand is they use YaRN and scale up the RoPE base

**Comment by BoxiangW** [comment]
@tjoymeed Thanks for the questions! I think we can close this issue and try open some other specific feature requests.

**Comment by tjoymeed** [comment]
okay thanks!
I read this paper: Qwen 2.5: A Comprehensive Review of the
 Leading Resource-Efficient LLM with potentioal to
 Surpass All Competitors

which mentioned about all those features for Qwen 2.5.

So could you please make a list of the architectural characteristics of Qwen 2.5 to replicate, as feature requests?
